php/php.project/src/org/netbeans/modules/php/project/ComputeTestMethodAnnotations.java

```diff
@@ -0,0 +1,190 @@
+        }
+        usagesCount++;
+    }
+
+    public void unregister() {
+        usagesCount--;
+        if (usagesCount == 0) {
+            EditorRegistry.removePropertyChangeListener(this);
+            isRegistered = false;
+        }
+    }
+
+    @Override
+    public void propertyChange(PropertyChangeEvent event) {
+        JTextComponent textComponent = EditorRegistry.lastFocusedComponent();
+        if (textComponent != null) {
+            Document document = textComponent.getDocument();
+            if (document != null) {
+                FileObject fileObject = NbEditorUtilities.getFileObject(document);
+                if (fileObject != null) {
+                    if (FileUtils.isPhpFile(fileObject)) {
+                        if (event.getPropertyName().equals(EditorRegistry.FOCUS_GAINED_PROPERTY)) {
+                            handleFileChange(document);
+                            document.addDocumentListener(ComputeTestMethodAnnotations.this);
+                        }
+
+                        if (event.getPropertyName().equals(EditorRegistry.FOCUS_LOST_PROPERTY)) {
+                            document.removeDocumentListener(this);
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    @Override
+    public void insertUpdate(DocumentEvent event) {
+        handleFileChange(event.getDocument());
+    }
+
+    @Override
+    public void removeUpdate(DocumentEvent event) {
+        handleFileChange(event.getDocument());
+    }
+
+    @Override
+    public void changedUpdate(DocumentEvent event) {
+        handleFileChange(event.getDocument());
+    }
+
+    private void handleFileChange(Document doc) {
+        handledDocument = doc;
+        task.schedule(500);
+    }
+
+    @Override
+    public void run() {
+        if (handledDocument != null) {
+            List<TestMethod> testMethods = new ArrayList<>();
+            try {
+                List<TestMethod> methods = getTestMethods(handledDocument);
+
+                if (methods != null) {
+                    testMethods.addAll(methods);
+                }
+            } catch (BadLocationException ex) {
+                Exceptions.printStackTrace(ex);
+            }
+
+            TestMethodController.setTestMethods(handledDocument, Collections.emptyList());
+            TestMethodController.setTestMethods(handledDocument, testMethods);
+        }
+    }
+
+    public List<TestMethod> getTestMethods(Document document) throws BadLocationException {
+        FileObject fileObject = NbEditorUtilities.getFileObject(document);
+        if (fileObject != null) {
+            PhpProject project = PhpProjectUtils.getPhpProject(fileObject);
+            assert project != null;
+            PhpModule phpModule = project.getPhpModule();
+            for (PhpTestingProvider testingProvider : project.getTestingProviders()) {
+                if (testingProvider.isTestFile(phpModule, fileObject)) {
```

Comment:
Nitpick: we are always iterating over all the providers; to improve performance, we should quit this iteration once we find the proper one for this file.

Reply:
Or, is it possible to have more providers in one project? I am not sure, but I think so. Please, confirm, thanks.